src/platform/Darwin/DnssdImpl.cpp

```diff
@@ -162,17 +164,108 @@ MdnsContexts MdnsContexts::sInstance;
+                           DNSServiceRegisterReply callback, RegisterContext * sdCtx)
+{
+    auto err = DNSServiceRegister(&sdRef, flags, interfaceId, name, type, domain, hostname, ntohs(port), txtLen, txtRecord,
+                                  callback, sdCtx);
+    ChipLogProgress(Discovery, "Registering service %s on port %u with type: %s on interface id: %u hostname %s error: %d" PRIu32,
+                    name, port, type, interfaceId, hostname, err);
+    VerifyOrReturnError(kDNSServiceErr_NoError == err, sdCtx->Finalize(err));
+    return MdnsContexts::GetInstance().Add(sdCtx, sdRef);
+}
+
+CHIP_ERROR RegisterRecord(RegisterRecordContext * registerRecordCtx, const char * type, uint32_t interfaceId, const char * hostname)
+{
+    DNSServiceRef sdRef;
+    auto err = DNSServiceCreateConnection(&sdRef);
+    VerifyOrReturnError((kDNSServiceErr_NoError == err || sdRef != nullptr), registerRecordCtx->Finalize(err));
+
+    auto error = MdnsContexts::GetInstance().Add(registerRecordCtx, sdRef);
+    VerifyOrReturnError(CHIP_NO_ERROR == error, error);
+    // Get all the interface addresses
+    struct ifaddrs * ifaddr;
+    auto ret = getifaddrs(&ifaddr);
+    if (ret)
+    {
+        ChipLogError(Discovery, "Getting interface addresses failed %d.", ret);
+        freeifaddrs(ifaddr);
+    }
+    VerifyOrReturnError(ret == 0, registerRecordCtx->Finalize(ret));
+
+    // Call DNSServiceRegisterRecord for each interface if interface id is of type any so we can register the address for each
+    // interface Otherwise just register record for the given interface id
+    bool isInterfaceIndexAny = (interfaceId == kDNSServiceInterfaceIndexAny);
+    for (struct ifaddrs * ifa = ifaddr; ifa != NULL; ifa = ifa->ifa_next)
+    {
+        if (ifa->ifa_addr->sa_family == AF_INET)
+        {
+            struct sockaddr_in * inaddr = (struct sockaddr_in *) ifa->ifa_addr;
+            bool shouldRegisterRecord =
+                (isInterfaceIndexAny || (!isInterfaceIndexAny && (if_nametoindex(ifa->ifa_name) == interfaceId)));
+            if (inaddr && shouldRegisterRecord)
+            {
+                ChipLogDetail(Discovery, "Registering Record for hostname %s interface %d", hostname,
+                              if_nametoindex(ifa->ifa_name));
+                DNSRecordRef dnsRecordRef;
+                err = DNSServiceRegisterRecord(sdRef, &dnsRecordRef, kRegisterRecordFlags, if_nametoindex(ifa->ifa_name), hostname,
+                                               kDNSServiceType_A, kDNSServiceClass_IN, 4, &inaddr->sin_addr, 0, OnRegisterRecord,
+                                               registerRecordCtx);
+                if (err || dnsRecordRef == nullptr)
+                {
+                    freeifaddrs(ifaddr);
+                }
+                VerifyOrReturnError(kDNSServiceErr_NoError == err, registerRecordCtx->Finalize(err));
```

Comment:
Why are we not checking dnsRecordRef here?  Or why are we checking it above when deciding whether to freeifaddrs?

More generally, if we are doing an error check anyway to do freeifaddrs, why are we bothering with a separate VerifyOrReturnError instead of just doing our cleanup and returning in the existing error check block?

Also, if this fails for some interface and `isInterfaceIndexAny`, should we really give up, or keep trying on other interfaces?

Reply:
let me re-evaluate this part. always better to return from the error check. we should continue if we fail for an interface.